File: frontend/src/api/roleTemplate.ts
import { apiGet, apiPost } from './client'

export interface RoleTemplate {
  id: number
  key: string
  name: string
  category: string | null
  description: string | null
  execution_mode: string
  default_role_name: string | null
  default_model_id: number | null
  default_skill_ids: number[]
  default_tool_ids: number[]
  default_code_policy: CodeExecutionPolicy | null
  is_builtin: boolean
  enabled: boolean
}

export interface CodeExecutionPolicy {
  max_iterations: number
  allow_file_write: boolean
  run_lint: boolean
  run_build: boolean
  run_unit_tests: boolean
  require_plan_first: boolean
  require_integration_tests: boolean
  allow_repo_read: boolean
  allow_repo_write: boolean
  allow_auto_commit: boolean
  allow_auto_pr: boolean
  stop_on_critical_error: boolean
  fallback_to_human: boolean
}

export interface CodeIteration {
  id: number
  code_task_id: number
  iteration_no: number
  status: string
  plan_summary: string | null
  changed_files: string | null
  validation_lint: string | null
  validation_build: string | null
  validation_unit_tests: string | null
}

interface RoleTemplateListResponse {
  items: RoleTemplate[]
}

interface CodeIterationListResponse {
  items: CodeIteration[]
}

export function listRoleTemplates(): Promise<RoleTemplateListResponse> {
  return apiGet<RoleTemplateListResponse>('/role-templates')
}

export function getRoleTemplate(id: number): Promise<RoleTemplate> {
  return apiGet<RoleTemplate>(`/role-templates/${id}`)
}

export function createAgentFromTemplate(projectId: number, data: { role_template_id: number; name?: string; description?: string; model_id?: number }): Promise<any> {
  return apiPost(`/projects/${projectId}/agents/from-template`, data)
}

export function listCodeIterationsByRun(runId: number): Promise<CodeIterationListResponse> {
  return apiGet<CodeIterationListResponse>(`/runs/${runId}/code-iterations`)
}

File: backend/app/schemas/role_template.py
from enum import Enum

from pydantic import BaseModel


class ExecutionModeEnum(str, Enum):
    normal_llm = "normal_llm"
    code_runtime = "code_runtime"
    analysis_runtime = "analysis_runtime"
    test_runtime = "test_runtime"


class CodeExecutionPolicySchema(BaseModel):
    max_iterations: int = 3
    allow_file_write: bool = False
    run_lint: bool = True
    run_build: bool = False
    run_unit_tests: bool = True
    require_plan_first: bool = True
    require_integration_tests: bool = False
    allow_repo_read: bool = True
    allow_repo_write: bool = False
    allow_auto_commit: bool = False
    allow_auto_pr: bool = False
    stop_on_critical_error: bool = True
    fallback_to_human: bool = True


class RoleTemplateListItem(BaseModel):
    id: int
    key: str
    name: str
    execution_mode: ExecutionModeEnum

    model_config = {"from_attributes": True}


class RoleTemplateResponse(BaseModel):
    id: int
    key: str
    name: str
    category: str | None = None
    description: str | None = None
    execution_mode: ExecutionModeEnum
    default_role_name: str | None = None
    default_model_id: int | None = None
    default_skill_ids: list[int] = []
    default_tool_ids: list[int] = []
    default_code_policy: CodeExecutionPolicySchema | None = None
    is_builtin: bool = False
    enabled: bool = True

    model_config = {"from_attributes": True}


class RoleTemplateListResponse(BaseModel):
    items: list[RoleTemplateListItem]
